planner/core/exhaust_physical_plans.go

```diff
@@ -716,6 +716,21 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
 		// On this way, it's double read case.
 		ts := PhysicalTableScan{Columns: ds.Columns, Table: is.Table, TableAsName: ds.TableAsName}.Init(ds.ctx, ds.blockOffset)
 		ts.SetSchema(is.dataSourceSchema)
+		// If inner cop task need keep order, the extraHandleCol should be set.
+		if cop.keepOrder {
+			cop.extraHandleCol, cop.doubleReadNeedProj = ts.appendExtraHandleCol(ds)
+			schemaColumns := make([]*expression.Column, len(is.dataSourceSchema.Columns))
+			copy(schemaColumns, is.dataSourceSchema.Columns)
+			ts.schema = ts.schema.Clone()
+			ts.schema.Columns = schemaColumns
+			// If the double read need projection, the datasource schema will append a handle column during appendExtraHandleCol,
+			// So we need to delete the extra column in schema to avoid wrong plan.
+			if cop.doubleReadNeedProj {
+				defer func() {
+					ds.schema.Columns = ds.schema.Columns[:len(ds.schema.Columns)-1]
```

Comment:
where did you change DataSource's schema?
The added projection won't affect your execution logic?

Reply:
1. DataSource's schema is changed in `ts.appendExtraHandleCol(ds)`
2. It does affect execution. So I will change the executing logic in PR of index merge join.

Reply:
```
// If there is a table reader which needs to keep order, we should append a pk to table scan.
func (ts *PhysicalTableScan) appendExtraHandleCol(ds *DataSource) {
	if len(ds.schema.TblID2Handle) > 0 {
		return
	}
	pkInfo := model.NewExtraHandleColInfo()
	ts.Columns = append(ts.Columns, pkInfo)
	handleCol := ds.newExtraHandleSchemaCol()
	ts.schema.Append(handleCol)
	ts.schema.TblID2Handle[ds.tableInfo.ID] = []*expression.Column{handleCol}
}
```

```
func (ds *DataSource) newExtraHandleSchemaCol() *expression.Column {
	return &expression.Column{
		DBName:   ds.DBName,
		TblName:  ds.tableInfo.Name,
		ColName:  model.ExtraHandleName,
		RetType:  types.NewFieldType(mysql.TypeLonglong),
		UniqueID: ds.ctx.GetSessionVars().AllocPlanColumnID(),
		ID:       model.ExtraHandleID,
	}
}
```

As you can see, datasource's schema is not changed. And it should not be changed.

Reply:
`ts.schema.Append(handleCol)` this line will change.

Reply:
`ts.schema` and `ds.schema` are same pointer here.

Reply:
```
// If there is a table reader which needs to keep order, we should append a pk to table scan.
func (ts *PhysicalTableScan) appendExtraHandleCol(ds *DataSource) {
	if len(ds.schema.TblID2Handle) > 0 {
		return
	}
	pkInfo := model.NewExtraHandleColInfo()
	ts.Columns = append(ts.Columns, pkInfo)
	handleCol := ds.newExtraHandleSchemaCol()
	ts.schema.Append(handleCol)
	ts.schema.TblID2Handle[ds.tableInfo.ID] = []*expression.Column{handleCol}
}
```

```
func (ds *DataSource) newExtraHandleSchemaCol() *expression.Column {
	return &expression.Column{
		DBName:   ds.DBName,
		TblName:  ds.tableInfo.Name,
		ColName:  model.ExtraHandleName,
		RetType:  types.NewFieldType(mysql.TypeLonglong),
		UniqueID: ds.ctx.GetSessionVars().AllocPlanColumnID(),
		ID:       model.ExtraHandleID,
	}
}
```

As you can see, datasource's schema is not changed. And it should not be changed.

Reply:
ok then, you should first let ts.schema and ds.schema not be the same thing. Then don't change ds.schema.

Reply:
Yeah. That's a better way.